### PR TITLE
Add Emoji/FA Autocompletion

### PR DESCRIPTION
**Pull request by DerMolly**
### Component/Part
markdown editor

### Description
This PR adds autocompletion for emojis and fork-awesome.

### Steps

- [x] added implementation
- [x] added / updated tests
- [ ] added / updated documentation
- [x] extended changelog

### Related Issue(s)
#316

### CODE REVIEW

File: src/components/editor/editor-window/autocomplete/autocomplete.ts [hunk 1]
```diff
@@ -16,13 +17,27 @@ export const emojiStrategy = (editor: Editor): StrategyProps => ({
...
+      // noinspection CheckTagEmptyBody
+      replacement = `<i class="fa ${(value as CustomEmoji).name}"></i> ${(value as BaseEmoji).name}`
+    }
+    return replacement
   },
   replace: (value) => {
-    return `$1${(value as BaseEmoji).colons}`
+    let replacement = ''
+    if ((value as BaseEmoji).native) {
+      replacement = `${(value as BaseEmoji).native}`
```
Comment (ErikMichelson) [inline src/components/editor/editor-window/autocomplete/autocomplete.ts]: Should autocomplete input the native emoji here instead of the shortcode variant (`colons`)?

Comment (ErikMichelson) [reply src/components/editor/editor-window/autocomplete/autocomplete.ts]: Nevermind, just saw #397

File: src/components/editor/editor-window/editor-window.tsx [hunk 2]
```diff
@@ -1,4 +1,5 @@
-import { Editor } from 'codemirror'
+import * as codemirror from 'codemirror'
```
Comment (DerMolly) [inline src/components/editor/editor-window/editor-window.tsx:1]: Why is this necessary?

File: src/components/editor/editor-window/editor-window.tsx [hunk 3]
```diff
@@ -14,22 +15,104 @@ import 'codemirror/addon/search/match-highlighter'
...
 import { ToolBar } from './tool-bar/tool-bar'
+import { addEmoji } from './tool-bar/utils'
+import data from 'emoji-mart/data/twitter.json'
 
 export interface EditorWindowProps {
   onContentChange: (content: string) => void
   content: string
 }
 
+interface findWOrdAtCursorResponse {
```
Comment (DerMolly) [inline src/components/editor/editor-window/editor-window.tsx:35]: Typo

File: src/components/editor/editor-window/editor-window.tsx [hunk 4]
```diff
@@ -14,22 +15,104 @@ import 'codemirror/addon/search/match-highlighter'
...
+  start: number,
+  end: number,
+  text: string
+}
+
+const allowedCharsInEmojiCodeRegex = /(:|\w)/
```
Comment (DerMolly) [inline src/components/editor/editor-window/editor-window.tsx:41]: Does \w contain '-'?

File: src/components/editor/editor-window/editor-window.tsx [hunk 5]
```diff
@@ -14,22 +15,104 @@ import 'codemirror/addon/search/match-highlighter'
...
+
+const findWordAtCursor = (cm: Editor):findWOrdAtCursorResponse => {
```
Comment (DerMolly) [inline src/components/editor/editor-window/editor-window.tsx:43]:
```suggestion
const findWordAtCursor = (editor: Editor): findWOrdAtCursorResponse => {
```

The Rest would need to change accordingly…

File: src/components/editor/editor-window/editor-window.tsx [hunk 6]
```diff
@@ -14,22 +15,104 @@ import 'codemirror/addon/search/match-highlighter'
...
+  return {
+    text: line.slice(start, end).toLowerCase(),
+    start: start,
+    end: end
+  }
+}
+
+const emojiWordRegex = /:(\w+)/
+
+const emojiHints = (cm: Editor) => {
```
Comment (DerMolly) [inline src/components/editor/editor-window/editor-window.tsx:64]: cm should be called editor

File: src/components/editor/editor-window/editor-window.tsx [hunk 7]
```diff
@@ -14,22 +15,104 @@ import 'codemirror/addon/search/match-highlighter'
...
+    if (searchresult === null) {
+      resolve(null)
+      return
+    }
+    const term = searchresult[1]
+    if (!term) {
+      resolve(null)
+      return
+    }
+    const search = emojiIndex.search(term)
```
Comment (DerMolly) [inline src/components/editor/editor-window/editor-window.tsx:77]: What about the Fork-Awesome Icons?

File: src/components/editor/editor-window/editor-window.tsx [hunk 8]
```diff
@@ -14,22 +15,104 @@ import 'codemirror/addon/search/match-highlighter'
...
+    const cursor = cm.getCursor()
+    resolve(search ? {
+      list: search.map(x => {
+        return {
+          displayText: `${addEmoji(x)} - ${x.colons ?? ''}`,
```
Comment (DerMolly) [inline src/components/editor/editor-window/editor-window.tsx:82]: What about the name for the fork-awesome icons?

File: src/components/editor/editor-window/editor-window.tsx (hunk 4 above)
Comment (SuperSandro2000) [reply src/components/editor/editor-window/editor-window.tsx:41]: Nope, only _.

**Comment by DerMolly** [comment]
Now the autocompletion also supports fork-awesome icons, but the both types are not correctly rendered in the overlay.
Bildschirmfoto von 2020-08-12 20-46-34
Bildschirmfoto von 2020-08-12 20-46-38

File: src/utils/emoji.ts [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+  return ''
+}
+
+export const getEmojiShortCode = (emoji: EmojiData):string => {
+  if ((emoji as BaseEmoji).colons) {
+    return (emoji as BaseEmoji).colons
+  } else if ((emoji as CustomEmoji).imageUrl) {
+    // noinspection CheckTagEmptyBody
+    return `:${(emoji as CustomEmoji).name}:`
+  }
```
Comment (ErikMichelson) [inline src/utils/emoji.ts:19]: `.colons` does exist on BaseEmoji as well as on CustomEmoji. You'll probably never end up  in the `else if` branch.
To distinguish between BaseEmoji and CustomEmoji, you have to test for `.native` and `.imageUrl`.

Comment (mrdrogdrog) [reply src/utils/emoji.ts:19]: It doesn't matter if it exists, only if it is filled.

Comment (ErikMichelson) [reply src/utils/emoji.ts]: Resolved in c6c72bc5

File: src/components/editor/editor-window/editor-window.tsx (hunk 4 above)
Comment (mrdrogdrog) [reply src/components/editor/editor-window/editor-window.tsx]: fixed

File: src/components/editor/editor-window/tool-bar/utils.test.ts [hunk 10]
```diff
@@ -1739,9 +1739,10 @@ describe('test addEmoji with native emoji', () => {
   const { cursor, firstLine, multiline, multilineOffset } = buildRanges()
   const textFirstLine = testContent.split('\n')[0]
   // noinspection CheckTagEmptyBody
```
Comment (DerMolly) [inline src/components/editor/editor-window/tool-bar/utils.test.ts:1741]: this can be removed

File: src/components/editor/editor-window/tool-bar/utils.ts [hunk 11]
```diff
@@ -24,13 +25,7 @@ export const addComment = (editor: Editor): void => changeLines(editor, line =>
...
 export const addEmoji = (emoji: EmojiData, editor: Editor): void => {
-  let replacement = ''
-  if ((emoji as BaseEmoji).colons) {
-    replacement = (emoji as BaseEmoji).colons
-  } else if ((emoji as CustomEmoji).imageUrl) {
-    // noinspection CheckTagEmptyBody
-    replacement = `<i class="fa ${(emoji as CustomEmoji).name}"></i>`
-  }
+  const replacement = getEmojiShortCode(emoji)
   insertAtCursor(editor, replacement)
```
Comment (DerMolly) [inline src/components/editor/editor-window/tool-bar/utils.ts:29]:
```suggestion
  insertAtCursor(editor, getEmojiShortCode(emoji))
```

File: src/components/editor/editor-window/editor-window.tsx [hunk 12]
```diff
@@ -10,14 +10,17 @@ import 'codemirror/addon/edit/matchbrackets'
 import 'codemirror/addon/edit/matchtags'
 import 'codemirror/addon/fold/foldcode'
 import 'codemirror/addon/fold/foldgutter'
+import 'codemirror/addon/hint/show-hint'
 import 'codemirror/addon/search/match-highlighter'
 import 'codemirror/addon/selection/active-line'
 import 'codemirror/keymap/sublime.js'
 import 'codemirror/mode/gfm/gfm.js'
-import React, { useState } from 'react'
+
```
Comment (DerMolly) [inline src/components/editor/editor-window/editor-window.tsx:18]: is this empty row necessary?

Comment (mrdrogdrog) [reply src/components/editor/editor-window/editor-window.tsx:18]: no
